# Partie recommandation
Réalisée par MUNOZ Matéo.

In [788]:
import pandas as pd
import sklearn as sk

In [789]:
data_product = pd.read_csv('../datasets/olist_products_dataset.csv')
data_order_item = pd.read_csv('../datasets/olist_order_items_dataset.csv')
data_customer = pd.read_csv('../datasets/olist_customers_dataset.csv')
data_review = pd.read_csv('../datasets/olist_order_reviews_dataset.csv')
data_order = pd.read_csv('../datasets/olist_orders_dataset.csv')
data_order_payments = pd.read_csv("../datasets/olist_order_payments_dataset.csv")


### Jeu de données order

Pour ce jeu, on va drop les colonnes qui ne nous serviront pas dans notre analyse. Notamment, on drop les colonnes qui traitent du statut de livraison.

In [790]:
print(len(data_order))
data_order.drop_duplicates(inplace=True)
print(len(data_order))

99441
99441


In [791]:
print(data_order.isnull().sum())

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


In [792]:
data_order.drop(["order_approved_at", "order_delivered_carrier_date", "order_delivered_customer_date", "order_estimated_delivery_date"], axis=1, inplace=True)
data_order.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp  
0    delivered      2017-10-02 10:56:33  
1    delivered      2018-07-24 20:41:37  
2    delivered      2018-08-08 08:38:49  
3    delivered      2017-11-18 19:28:06  
4    delivered      2018-02-13 21:18:39

On va aussi transformer les timestamp en datetime, car ce sont actuellement des chaînes de charactères

In [793]:
data_order["order_purchase_timestamp"] = pd.to_datetime(data_order["order_purchase_timestamp"])
data_order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   order_id                  99441 non-null  object        
 1   customer_id               99441 non-null  object        
 2   order_status              99441 non-null  object        
 3   order_purchase_timestamp  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(3)
memory usage: 3.0+ MB


### Jeu de données customer

In [794]:
print(len(data_customer))
data_customer.drop_duplicates(inplace=True)
print(len(data_customer))

data_customer.isnull().sum()

99441
99441


customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

### Jeu de données order_item

Pour ce jeu de données, on va aussi drop la colonne qui concerne les livraisons.

In [795]:
print(len(data_order_item))
data_order_item.drop_duplicates(inplace=True)
print(len(data_order_item))

print()

print(data_order_item.isnull().sum())

print()

data_order_item.drop("shipping_limit_date", axis=1, inplace=True)
data_order_item.head()

112650
112650

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64



order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id   price  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   58.90   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36  239.90   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d  199.00   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   12.99   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87  199.90   

   freight_value  
0          13.29  
1          19.93  
2          17.87  
3          12.79  
4          18.14

### Jeu de données order_payments

In [796]:
print(len(data_order_payments))
data_order_payments.drop_duplicates(inplace=True)
print(len(data_order_payments))

print()

print(data_order_payments.isnull().sum())

103886
103886

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64


### Jeu de données product

In [797]:
print(len(data_product))
data_product.drop_duplicates(inplace=True)
print(len(data_product))

data_product.isnull().sum()

32951
32951


product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [798]:
data_product = data_product[["product_id", "product_category_name"]]
data_product.head(5)

product_id  product_category_name
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer
3  cef67bcfe19066a932b7673e239eb23d                  bebes
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas

In [799]:
translate_df = pd.read_csv("../datasets/product_category_name_translation.csv")

translate_dict = dict()
for record in translate_df.to_records():
    translate_dict[record[1]] = record[2]

translate_dict

data_product["product_category_name"] = data_product["product_category_name"].replace(translate_dict)
data_product.head(10)

product_id product_category_name
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumery
1  3aa071139cb16b67ca9e5dea641aaa2f                   art
2  96bd76ec8810374ed1b65e291975717f        sports_leisure
3  cef67bcfe19066a932b7673e239eb23d                  baby
4  9dc1a7de274444849c219cff195d0b71            housewares
5  41d3672d4792049fa1779bb35283ed13   musical_instruments
6  732bd381ad09e530fe0a5f457d81becb            cool_stuff
7  2548af3e6e77a690cf3eb6368e9ab61e       furniture_decor
8  37cc742be07708b53a98702e77a21a02       home_appliances
9  8c92109888e8cdf9d66dc7e463025574                  toys

### Jeu de données order_review

In [800]:
print(len(data_review))
data_review.drop_duplicates(inplace=True)
print(len(data_review))

99224
99224


In [801]:
data_review.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

Beaucoup de review n'ont pas de commentaire et de titre, on va donc drop les colonnes `review_comment_title` et `review_comment_message`.
On va aussi drop les colonnes de timestamp car elles ne sont pas utiles à notre système de recommandations.

In [802]:
data_review.drop("review_comment_title", axis=1, inplace=True)
data_review.drop("review_comment_message", axis=1, inplace=True)
data_review.drop("review_creation_date", axis=1, inplace=True)
data_review.drop("review_answer_timestamp", axis=1, inplace=True)
data_review.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score  
0             4  
1             5  
2             5  
3             5  
4             5

### Fusion des données obtenues

In [803]:
final_data = data_order.merge(data_order_item, on="order_id", how="left")
final_data = final_data.merge(data_order_payments, on="order_id", how="outer")
final_data = final_data.merge(data_product, on="product_id", how="outer")
final_data = final_data.merge(data_customer, on="customer_id", how="outer")
final_data = final_data.merge(data_review, on="order_id", how="outer")

In [804]:
final_data.head()

order_id                       customer_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  3ce436f183e68e07877b285a838db11a   
1  00018f77f2f0320c557190d7a144bdd3  f6dd3ec061db4e3987629fe6b26e5cce   
2  000229ec398224ef6ca0657da4fc703e  6489ae5e4333f3693df5ad4372dab6d3   
3  00024acbcdf0a6daa1e931b038114c75  d4eb9395c8c0431ee92fce09860c5a06   
4  00042b26cf59d7ce69dfabb4e55b4fd9  58dbd0b2d70206bf40e62cd34e84d795   

  order_status order_purchase_timestamp  order_item_id  \
0    delivered      2017-09-13 08:59:02            1.0   
1    delivered      2017-04-26 10:53:06            1.0   
2    delivered      2018-01-14 14:33:31            1.0   
3    delivered      2018-08-08 10:00:35            1.0   
4    delivered      2017-02-04 13:57:51            1.0   

                         product_id                         seller_id   price  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   58.90   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36  239.90   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d  199.00   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   12.99   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87  199.90   

   freight_value  payment_sequential payment_type  payment_installments  \
0          13.29                 1.0  credit_card                   2.0   
1          19.93                 1.0  credit_card                   3.0   
2          17.87                 1.0  credit_card                   5.0   
3          12.79                 1.0  credit_card                   2.0   
4          18.14                 1.0  credit_card                   3.0   

   payment_value product_category_name                customer_unique_id  \
0          72.19            cool_stuff  871766c5855e863f6eccc05f988b23cb   
1         259.83              pet_shop  eb28e67c4c0b83846050ddfb8a35d051   
2         216.87       furniture_decor  3818d81c6709e39d06b2738a8d3a2474   
3          25.78             perfumery  af861d436cfc08b2c2ddefd0ba074622   
4         218.04          garden_tools  64b576fb70d441e8f1b2d7d446e483c5   

   customer_zip_code_prefix          customer_city customer_state  \
0                     28013  campos dos goytacazes             RJ   
1                     15775        santa fe do sul             SP   
2                     35661          para de minas             MG   
3                     12952                atibaia             SP   
4                     13226        varzea paulista             SP   

                          review_id  review_score  
0  97ca439bc427b48bc1cd7177abe71365           5.0  
1  7b07bacd811c4117b742569b04ce3580           4.0  
2  0c5b33dea94867d1ac402749e5438e8b           5.0  
3  f4028d019cb58564807486a6aaf33817           4.0  
4  940144190dcba6351888cafa43f3a3a5           5.0

## Explication des données obtenues

On va effectuer une série d'analyse pour expliquer les données que l'on a obtenue.

### Review score

In [805]:
import plotly.express as px

px.bar(final_data['review_score'].value_counts())

In [806]:
product_review_counts = final_data['product_category_name'].value_counts().reset_index()
product_review_counts.columns = ['product_category_name', 'review_count']

fig = px.bar(
    product_review_counts,
    x='product_category_name',
    y='review_count',
    title='Nombre de notes par catégorie de produit',
    labels={'product_category_name': 'Catégorie de produit', 'review_count': 'Nombre de notes'},
)
fig.update_layout(xaxis={'categoryorder':'total descending'}, xaxis_tickangle=-45)
fig.show()

In [807]:
product_review_counts = final_data.groupby('product_id')['review_score'].size()

average_reviews_per_product = product_review_counts.mean()

print(f"Nombre moyen de reviews par produit : {average_reviews_per_product:.2f}")


Nombre moyen de reviews par produit : 3.59


In [808]:
user_review_counts = final_data.groupby('customer_unique_id')['review_score'].count()
average_reviews_per_user = user_review_counts.mean()

print(f"Le nombre moyen de notes par utilisateur est : {average_reviews_per_user:.2f}")

Le nombre moyen de notes par utilisateur est : 1.23


In [809]:
import pandas as pd
import plotly.express as px

user_review_counts = final_data.groupby('customer_unique_id')['review_score'].count().reset_index()
user_review_counts.columns = ['customer_unique_id', 'review_count']

print(user_review_counts['review_count'].max())
bins = [1, 2, 3, 4, user_review_counts['review_count'].max()]
labels = ['1', '2', '3', '4+']

user_review_counts['review_category'] = pd.cut(user_review_counts['review_count'], bins=bins, labels=labels, right=False)

review_category_counts = user_review_counts['review_category'].value_counts().sort_index().reset_index()
review_category_counts.columns = ['review_category', 'user_count']

fig = px.bar(
    review_category_counts,
    x='review_category',
    y='user_count',
    title="Répartition des utilisateurs par nombre de notes données",
    labels={'review_category': 'Nombre de notes données', 'user_count': 'Nombre d\'utilisateurs'},
)

fig.show()


75


On observe que la grande majorité des notes sont de 5, et surtout que peut de notes se situent dans les valeurs intermédiaires (2 et 3).

On observe aussi que les notes ne sont pas très bien réparties entre toutes les catégories de produit, certaines catégories récoltent beaucoup de notes tandis que d'autre en on peu ou presque pas. Aussi le nombre de note moyenne par produit est de 3.59.

Finalement, on voit que le nombre moyen de notes par utilisateur est de 1.23, et qu'on a effectivement un nombre écrasant d'utilisateur qui ne donne qu'une seul note.

### Autre détail sur les données

In [810]:
print(f'Il y a {len(pd.unique(final_data["product_id"]))} produit unique.')
print(f'Il y a {len(pd.unique(final_data["customer_unique_id"]))} client unique.')
print(f'Il y a {len(pd.unique(final_data["seller_id"]))} vendeur unique.')
print(f'Il y a {len(pd.unique(final_data["product_category_name"]))} catégorie de produit unique.')


Il y a 32952 produit unique.
Il y a 96096 client unique.
Il y a 3096 vendeur unique.
Il y a 74 catégorie de produit unique.


### Système de recommendation
On va commencer par retirer les données qui ne nous servent pas pour faire la recommendation, puis on va transformer celle dont on a besoin en nombre (comme les identifiants par exemple qui sont des strings). On va aussi retirer tous les utilisateur qui n'ont pas de review du set, et on va retirer tous les produits qui ont un nombre de notes inférieur à la moyenne de 3.59.

In [811]:
user_review_counts = final_data['customer_unique_id'].value_counts()

active_users = user_review_counts[user_review_counts > 1].index
final_data = final_data[final_data['customer_unique_id'].isin(active_users)]

product_avg_reviews_nb = final_data.groupby('product_id')['review_score'].count().reset_index()
product_avg_reviews_nb.columns = ['product_id', 'avg_review_nb']

filtered_products = product_avg_reviews_nb[product_avg_reviews_nb['avg_review_nb'] >= 2]
final_data = final_data[final_data['product_id'].isin(filtered_products['product_id'])]

num_users = final_data['customer_unique_id'].nunique()
num_product = final_data['product_id'].nunique()

print(f"Nombre d'utilisateurs après filtrage : {num_users}")
print(f"Nombre de produits après filtrage : {num_product}")

Nombre d'utilisateurs après filtrage : 13380
Nombre de produits après filtrage : 7361


In [812]:
from sklearn.preprocessing import LabelEncoder

customer_encoder = LabelEncoder()
product_encoder = LabelEncoder()

data_ratings = final_data[['customer_unique_id', 'product_id', 'review_score', 'product_category_name']].copy()
data_ratings.dropna(inplace=True)

data_ratings['customer_unique_id'] = customer_encoder.fit_transform(data_ratings['customer_unique_id'])
data_ratings['product_id'] = product_encoder.fit_transform(data_ratings['product_id'])

product_id_mapping = dict(zip(product_encoder.classes_, product_encoder.transform(product_encoder.classes_)))

print(f'Nombres de lignes du tableau final : {data_ratings.shape[0]}')

data_ratings.head()

Nombres de lignes du tableau final : 31958


customer_unique_id  product_id  review_score  product_category_name
13                8109        1555           5.0           garden_tools
14                8109        1555           5.0           garden_tools
39                2674        1837           5.0           garden_tools
40                2674        1837           5.0           garden_tools
41                7834        7170           1.0  computers_accessories

On va normaliser les scores entre 0 et 1.

In [813]:
from sklearn.preprocessing import MinMaxScaler

score_normalizer = MinMaxScaler()

data_ratings['review_score'] = score_normalizer.fit_transform(data_ratings[['review_score']])

data_ratings.head()

customer_unique_id  product_id  review_score  product_category_name
13                8109        1555           1.0           garden_tools
14                8109        1555           1.0           garden_tools
39                2674        1837           1.0           garden_tools
40                2674        1837           1.0           garden_tools
41                7834        7170           0.0  computers_accessories

On regarde la distribution des scores.

In [814]:
print(data_ratings['review_score'].describe())

count    31958.000000
mean         0.691603
std          0.387042
min          0.000000
25%          0.500000
50%          1.000000
75%          1.000000
max          1.000000
Name: review_score, dtype: float64


On remarque que dans la distribution, les notes maximales sont sûr-représentées, ce qui peut biaiser le modèle. On va donc centrer les données.

In [815]:
mean = data_ratings['review_score'].mean()
std = data_ratings['review_score'].std()
data_ratings['standardized_score'] = (data_ratings['review_score'] - mean) / std

print(data_ratings['standardized_score'].describe())

count    3.195800e+04
mean     6.781261e-17
std      1.000000e+00
min     -1.786895e+00
25%     -4.950447e-01
50%      7.968051e-01
75%      7.968051e-01
max      7.968051e-01
Name: standardized_score, dtype: float64


On obtient finalement les données suivantes :

In [816]:
data_ratings.drop(columns=['review_score'], inplace=True)
data_ratings.head()

customer_unique_id  product_id  product_category_name  standardized_score
13                8109        1555           garden_tools            0.796805
14                8109        1555           garden_tools            0.796805
39                2674        1837           garden_tools            0.796805
40                2674        1837           garden_tools            0.796805
41                7834        7170  computers_accessories           -1.786895

In [817]:
from math import inf
import joblib
import numpy as np
from surprise import SVD, Dataset, Reader, NMF, SVDpp
from surprise.model_selection import cross_validate

def found_best_model(factors: list, data, model_name, model_type: str):
    best_rmse = inf
    best_model = None
    for n in factors:
        if model_type == 'svd':
            model = SVD(n_factors=n)
        elif model_type == 'nmf':
            model = NMF(n_factors=n)
        elif model_type == 'svdpp':
            model = SVDpp(n_factors=n)
        
        results1 = cross_validate(model, data, measures=['RMSE'], cv=5, verbose=False, n_jobs=-1)
        
        mean_rmse1 = np.mean(results1['test_rmse'])
        
        print(f"n_factors={n}, Mean RMSE: {mean_rmse1}")
        
        if mean_rmse1 < best_rmse:
            best_rmse = mean_rmse1
            best_model = model

    joblib.dump(best_model, model_name + '.pkl')
    print("Le meilleur modèle a été enregistré avec un RMSE de :", best_rmse)
    

reader = Reader(rating_scale=(data_ratings['standardized_score'].min(), data_ratings['standardized_score'].max()))
data_ratings_surprise = Dataset.load_from_df(data_ratings.drop(columns=['product_category_name']), reader)

factors = [2, 8, 16, 32, 64, 128, 512]
found_best_model(factors, data_ratings_surprise, model_name='best_svd', model_type='svd')
    

n_factors=2, Mean RMSE: 0.8105436077463812
n_factors=8, Mean RMSE: 0.8011358796951779
n_factors=16, Mean RMSE: 0.789063511760714
n_factors=32, Mean RMSE: 0.7730119269757225
n_factors=64, Mean RMSE: 0.7389658975174378
n_factors=128, Mean RMSE: 0.6938176535295806
n_factors=512, Mean RMSE: 0.5855443698689098
Le meilleur modèle a été enregistré avec un RMSE de : 0.5855443698689098


In [ ]:
factors = [5, 10, 15, 20, 30, 40, 50, 100]
found_best_model(factors, data_ratings_surprise, model_name='best_nmf', model_type='nmf')

n_factors=5, Mean RMSE: 1.1837675089160191
n_factors=10, Mean RMSE: 1.1771530848229592
n_factors=15, Mean RMSE: 1.1771172021051026
n_factors=20, Mean RMSE: 1.1688912991665996
n_factors=30, Mean RMSE: 1.1593038832684555
n_factors=40, Mean RMSE: 1.1669626475506116
n_factors=50, Mean RMSE: 1.1682276225484725
n_factors=100, Mean RMSE: 1.16142549913152
Le meilleur modèle a été enregistré avec un RMSE de : 1.1593038832684555


In [821]:
from surprise.model_selection.search import GridSearchCV

def optimize_model_with_grid_search(model_type: str, param_grid: dict, data):
    if model_type == 'svd':
        model = SVD
    elif model_type == 'nmf':
        model = NMF
    else:
        raise ValueError("Type de modèle non reconnu. Utilisez 'svd' ou 'nmf'.")

    grid_search = GridSearchCV(
        model,
        param_grid=param_grid,
        measures=['rmse'],
        cv=5,
        n_jobs=-1
    )

    grid_search.fit(data)

    best_params = grid_search.best_params['rmse']
    best_rmse = grid_search.best_score['rmse']
    print(f"Meilleurs paramètres : {best_params}")
    print(f"Meilleur RMSE : {best_rmse:.4f}")

    best_model = grid_search.best_estimator['rmse']
    best_model.fit(data.build_full_trainset())

    joblib.dump(best_model, 'best_model.pkl')
    print("Meilleur modèle sauvegardé sous 'best_model.pkl'.")

    return {
        'best_params': best_params,
        'best_rmse': best_rmse,
        'best_model': best_model
    }

param_grid = {
    'n_factors': [10, 50, 100, 200, 400, 500, 1000],
    'n_epochs': [30, 60, 100, 200, 500],
    'lr_all': [0.005, 0.0001, 0.0005, 0.00001, 0.001, 0.0009]
}

results = optimize_model_with_grid_search('svd', param_grid, data_ratings_surprise)


Meilleurs paramètres : {'n_factors': 10, 'n_epochs': 500, 'lr_all': 0.005}
Meilleur RMSE : 0.5113
Meilleur modèle sauvegardé sous 'best_model.pkl'.


In [822]:
import joblib

trainset = data_ratings_surprise.build_full_trainset()

model = joblib.load('best_model.pkl')
model.fit(trainset)

feature_mat = model.qi

product_id = [trainset.to_raw_iid(x) for x in range(len(feature_mat))]

feature_df = pd.DataFrame(feature_mat, index=product_id)
feature_df['category'] = feature_df.index.map(lambda pid: data_ratings.loc[data_ratings['product_id'] == pid, 'product_category_name'].iloc[0])

feature_df.head()

0         1         2         3         4         5         6  \
1555 -0.372702 -0.137664  0.254503 -0.378636  0.383877 -0.071174 -0.507024   
1837 -0.027407  0.089561  0.037653  0.218072 -0.010408 -0.110452 -0.030754   
7170 -0.113446  0.168135 -0.183052  0.155341  0.335935 -0.173181 -0.165706   
308   0.069217 -0.106211  0.038839 -0.116468 -0.045639 -0.058777  0.092334   
6466  0.012066  0.233940 -0.170853  0.124338 -0.216406  0.257243 -0.310200   

             7         8         9               category  
1555  0.043251  0.111120 -0.037302           garden_tools  
1837 -0.147868 -0.108112 -0.005641           garden_tools  
7170 -0.011715  0.189469  0.145336  computers_accessories  
308  -0.050894  0.057810 -0.010358            electronics  
6466  0.683249  0.434487 -0.189476          health_beauty

In [825]:
from sklearn.discriminant_analysis import StandardScaler
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA

scaler = StandardScaler()
standardized_features = scaler.fit_transform(feature_df.drop(columns=['category']))

pca = PCA(n_components=5)

tsne = TSNE(2, perplexity=10)

pre_reduced_f = pca.fit_transform(standardized_features)
reduced_feat = tsne.fit_transform(pre_reduced_f)

In [826]:
tsne_df = pd.DataFrame(reduced_feat, index=product_id, columns=['Factor 1', 'Factor 2'])
tsne_df['category'] = feature_df['category']


px.scatter(tsne_df, 
    x='Factor 1', 
    y='Factor 2',
    hover_name=tsne_df['category'])

In [843]:
from surprise import Reader
from surprise import Dataset


def recommend_product_for_customer(model, customer_id, data_ratings, n_recommendations=5):
    reader = Reader()
    data_ratings_surprise = Dataset.load_from_df(data_ratings.drop(columns=['product_category_name']), reader)
    
    trainset = data_ratings_surprise.build_full_trainset()

    model.fit(trainset)

    rated_product_row = data_ratings[data_ratings['customer_unique_id'] == customer_id]
    
    all_product = data_ratings['product_id'].unique()
    rated_product = rated_product_row['product_id'].unique()
    product_to_predict = [product for product in all_product if product not in rated_product]
    
    predictions = [model.predict(customer_id, product) for product in product_to_predict]
    
    recommended_products = sorted(predictions, key=lambda x: x.est, reverse=True)[:n_recommendations]
    
    rated_prod_final = [(prod, score) for prod, score in rated_product_row[['product_id', 'standardized_score']].values.tolist()]

    results = {
        'Rated Product': rated_prod_final,
        'Recommended Product': [(int(pred.iid), pred.est) for pred in recommended_products]
    }

    return results

random_customer_id = data_ratings['customer_unique_id'].sample(1).iloc[0]

print(f"Client ID aléatoire : {random_customer_id}")
loaded_model = joblib.load('best_model.pkl')
results = recommend_product_for_customer(loaded_model, random_customer_id, data_ratings, 10)

print("Rated Product:")
for product, rating in results['Rated Product']:
    product = int(product)
    for k, v in product_id_mapping.items():
        if v == int(product):
            product_info = final_data[final_data['product_id'] == k]
            break
        
        
    print(f"{product} - {product_info['product_category_name'].values[0]}: {round((rating * 0.387042) + 0.691603, 2)}")

print("\nRecommended Product:")
for product, est_rating in results['Recommended Product']:
    for k, v in product_id_mapping.items():
        if v == int(product):
            product_info = final_data[final_data['product_id'] == k]
            break
    print(f"{product} - {product_info['product_category_name'].values[0]}: Estimated Rating = {est_rating:.2f}")

Client ID aléatoire : 6062
Rated Product:
5281 - health_beauty: 1.0
5339 - health_beauty: 1.0

Recommended Product:
5767 - housewares: Estimated Rating = 1.50
3344 - furniture_decor: Estimated Rating = 1.47
810 - home_construction: Estimated Rating = 1.39
1993 - pet_shop: Estimated Rating = 1.37
5117 - agro_industry_and_commerce: Estimated Rating = 1.36
499 - bed_bath_table: Estimated Rating = 1.35
3307 - furniture_decor: Estimated Rating = 1.35
2689 - sports_leisure: Estimated Rating = 1.34
809 - housewares: Estimated Rating = 1.30
1638 - sports_leisure: Estimated Rating = 1.29
